### PR TITLE
Improvements to stream_response_to_file()

### PR DESCRIPTION
**Pull request by rcoup**
Problem: API client library saving large responses via `requests_toolbelt.downloadutils.stream.stream_response_to_file()` wants to enable simple progress reporting via a callback.

Changes:

1. ~~add `callback` parameter to `stream_response_to_file()`, which is a function that takes two arguments `(bytes_read, total_bytes)`. The function will be called at least twice (at 0 bytes and at the end). `total_bytes` can be `None` if there's no `Content-Length` header on the response.~~
2. add optional `chunksize` parameter to `stream_response_to_file()`, ~~and increase the default chunk size from 512 bytes to 64KB. This matches the chunk size used in `dowloadutils.tee`. This should improve performance significantly for large responses.~~
3. extract the filename/header/file-object parsing into a separate function
4. add support for passing a directory to save into using the Content-Disposition filename (rather than always using the current working directory)
5. only use the last file-part of Content-Disposition headers: eg. `Content-Disposition: attachment; filename=/etc/passwd`
6. prevent overwriting existing files using arbitrary filenames from the server. This is obviously backwards incompatible for apps that expect to overwrite the same file every time.

* [x] Tests
* [x] Docs

### CODE REVIEW

File: requests_toolbelt/downloadutils/stream.py [hunk 1]
```diff
@@ -12,6 +12,7 @@
     r';\s*(%s|[^\s;=]+)\s*(?:=\s*(%s|[^;]+))?\s*' % (_QUOTED_STRING_RE,
                                                      _QUOTED_STRING_RE)
 )
+_DEFAULT_CHUNKSIZE = 65536
```
Comment (sigmavirus24) [inline requests_toolbelt/downloadutils/stream.py]: Why would you change the default chunksize from 512 to this?

File: requests_toolbelt/downloadutils/stream.py [hunk 2]
```diff
@@ -122,8 +138,19 @@ def stream_response_to_file(response, path=None):
             )
         fd = open(filename, 'wb')
 
-    for chunk in response.iter_content(chunk_size=512):
+    try:
+        bytes_total = int(response.headers.get('content-length', None))
+    except TypeError:
+        bytes_total = None  # not specified or invalid
```
Comment (sigmavirus24) [inline requests_toolbelt/downloadutils/stream.py]: If it's not specified than it was probably sent using `Transfer-Encoding: chunked` which is not invalid.

This should be done differently, *if* at all.

**Comment by sigmavirus24** [comment]
`stream_response_to_file` does not need to know about a callback. Here's why:

```py
class MonitoringFile(object):
    def __init__(self, filename, mode, total_bytes):
        self.filename = filename
        self.file = open(filename, mode)
        self.total_bytes = total_bytes
        self.bytes_written = 0

    def write(self, data):
        if not data:
            return
        self.bytes_written += len(data)
        self.file.write(data)

response = requests.get(url, stream=True)
total_bytes = get_download_length(response)
fd = MonitoringFile('my-streamed-file.txt', 'wb', total_bytes)
stream_response_to_file(response, fd)
```

**Comment by sigmavirus24** [comment]
That said, I'm fine with allowing the chunksize to be configurable, but not to changing it without good justification.

File: requests_toolbelt/downloadutils/stream.py (hunk 1 above)
Comment (rcoup) [inline requests_toolbelt/downloadutils/stream.py]: logic: 512 bytes is very low for most download where you care about streaming to a file, and `tee` already uses 64KB.

File: requests_toolbelt/downloadutils/stream.py (hunk 2 above)
Comment (rcoup) [inline requests_toolbelt/downloadutils/stream.py]: There's plenty of responses that (quite rightly) don't include a Content-Length. But yeah, can easily rework into an `if` statement and a separate `int()` cast .

**Comment by rcoup** [comment]
> stream_response_to_file does not need to know about a callback. Here's why:

Sure, it doesn't _need_ to, just trying to provide a helper. The wider use case for me is wrapping `stream_response_to_file()` as a method in another library, but keeping the useful functionality around being able to pass in:
* a file-like object
* a file path
* or have the filename parsed from `Content-Disposition`
And provide a simple progress callback option for a user of my library.

If we did the `MonitoringFile` approach then we'd need to reimplement the `path` handling at the outer level and after that there's not a lot left of it :)

File: requests_toolbelt/downloadutils/stream.py (hunk 1 above)
Comment (sigmavirus24) [inline requests_toolbelt/downloadutils/stream.py]: Except that tee has a much different use case. In this case, we're dealing with a streamed response from requests and there have been issues in the past where we found 512 to be a good chunksize. Making this configurable will allow you to default it to whatever you'd like in your wrapper.

**Comment by sigmavirus24** [comment]
> If we did the MonitoringFile approach then we'd need to reimplement the path handling at the outer level and after that there's not a lot left of it :)

Which suggests that we need to refactor that into a separate helper function in this module. That makes it more testable and also allows others to use it when implementing extra features.

I'm fairly certain this function doesn't need the ability to handle callbacks. @Lukasa, thoughts?

**Comment by rcoup** [comment]
Update - I have reworked into several changes as discussed above. New strategy is better?

Each is a separate commit, so hopefully straightforward to review that way.

File: requests_toolbelt/downloadutils/stream.py [hunk 3]
```diff
@@ -12,17 +14,72 @@
...
+    """
+    pre_opened = False
+    fd = None
+    filename = None
+    if path:
+        if isinstance(getattr(path, 'write', None), collections.Callable):
+            pre_opened = True
+            fd = path
+            filename = getattr(fd, 'name', None)
+        elif os.path.isdir(path):
```
Comment (sigmavirus24) [inline requests_toolbelt/downloadutils/stream.py]: Why not more simply:

```py
elif not os.path.isdir(path):
    # If the user specified a directory, let the 
    filename = path
    if os.path.exists(filename):
        raise
    if os.path.exists(filename):
        raise exc.StreamingError(...)
    fd = open(filename, 'wb')
```

That said, why are we now going to start allowing people to specify the path to a directory now?

**Comment by sigmavirus24** [comment]
I'm sorry. I wasn't very clear. I want a new *public* function that will simply return the filename to be used from a Response. This will allow you to call that first and then call `os.path.join(directory, filename)` and pass that to `stream_response_to_file`.

**Comment by rcoup** [comment]
> I want a...

No problem, will take another stab at it :)

> That said, why are we now going to start allowing people to specify the path to a directory now?

Downloading to arbitrary filenames in `cwd` seems like a bad idea to encourage. I mean, we can clearly document it but it's still not great. 

Downloading to a specified directory seems tidier, and doesn't require `os.chdir()` (which changes it for all threads in a MT app, and still isn't available as a context manager in stdlib).

Personally I'd suggest eventually deprecating `path=None` to avoid unexpected behaviour, and suggest users use `path=os.curdir` if that's what they _actually_ want.

**Comment by rcoup** [comment]
Revisiting an earlier point:

> stream_response_to_file does not need to know about a callback. Here's why:

Would you prefer including that example in the docs, or bundling a basic implementation ala `MultipartEncoderMonitor`?

**Comment by rcoup** [comment]
ok, implemented as `get_download_file_path()` (improved naming welcome), and moved more of the logic back into `stream_response_to_file()`. 

Re-included the dangerous pathname and existing file stuff as a separate commit.

**Comment by sigmavirus24** [comment]
Triggering a new merged build

**Comment by rcoup** [comment]
Fixed the pyflakes errors.

**Pull request merged**